xdem/terrain.py

```diff
@@ -342,71 +629,337 @@ def _make_attribute_from_coefs(
         # so we use a 1-d array and write in a 2-d array output
-        plancurv = np.where(
-            C[zt_g_idx] ** 2 + C[zt_h_idx] ** 2 == 0.0,
-            np.array([0.0]),
-            -2
-            * (
-                C[zt_d_idx] * C[zt_h_idx] ** 2
-                + C[zt_e_idx] * C[zt_g_idx] ** 2
-                - C[zt_f_idx] * C[zt_g_idx] * C[zt_h_idx]
+
+        if curv_method_id == 0:
+
+            # Geometric tangential curvature (normal contour curvature) following Krcho, 1983.
+            # tancurv = - (z_xx * z_y**2 - 2 * z_xy * z_x * z_y + z_yy * z_x**2) /
+            # ((z_x**2 + z_y**2) * sqrt(1 + z_x**2 + z_y**2))
+
+            tancurv = np.where(
+                C[z_x_idx] ** 2 + C[z_y_idx] ** 2 == 0.0,
+                np.array([0.0]),
+                -(
+                    C[z_xx_idx] * C[z_y_idx] ** 2
+                    - 2 * C[z_xy_idx] * C[z_x_idx] * C[z_y_idx]
+                    + C[z_yy_idx] * C[z_x_idx] ** 2
+                )
+                / ((C[z_x_idx] ** 2 + C[z_y_idx] ** 2) * np.sqrt(1 + C[z_x_idx] ** 2 + C[z_y_idx] ** 2)),
             )
-            / (C[zt_g_idx] ** 2 + C[zt_h_idx] ** 2)
-            * 100,
-        )
+
+            # convert from m-1 to 100 m-1
+            tancurv *= 100
+
+        if curv_method_id == 1:
+
+            # Directional derivative tangential curvature: 2nd contour derivative "plan curvature",
+            # following Zevenberg and Thorne, 1979.
+            # tancurv = -(z_xx * z_y**2 - 2 * z_xy * z_x * z_y + z_yy * z_x**2) / (z_x**2 + z_y**2)
+
+            tancurv = np.where(
+                C[z_x_idx] ** 2 + C[z_y_idx] ** 2 == 0.0,
+                np.array([0.0]),
+                -(
+                    C[z_xx_idx] * C[z_y_idx] ** 2
+                    - 2 * C[z_xy_idx] * C[z_x_idx] * C[z_y_idx]
+                    + C[z_yy_idx] * C[z_x_idx] ** 2
+                )
+                / (C[z_x_idx] ** 2 + C[z_y_idx] ** 2),
+            )
+
+            # convert from m-1 to 100 m-1
+            tancurv *= 100
+
+        if tancurv_idx != 99:
+            attrs[tancurv_idx] = tancurv
+
+    if make_planform_curvature and surface_fit_id in [1, 2]:
+
+        # # Completely flat surfaces need to be set to zero to avoid division by zero
+        # # Unfortunately np.where doesn't support scalar input or 0d-array for the Numba parallel case,
+        # # so we use a 1-d array and write in a 2-d array output
+
+        if curv_method_id in [0, 1]:
+
+            # Geometric planform curvature following Sobolevsky, 1932
+            # Planform derivation is the same in a geometric and directional derivative context
+            # (see Minár et al. 2020 following Jenčo, 1992)
+            # plancurv = - (z_xx * z_y**2 - 2 * z_xy * z_x * z_y + z_yy * z_x**2) / sqrt((z_x**2 + z_y**2)**3)
+
+            plancurv = np.where(
+                C[z_x_idx] ** 2 + C[z_y_idx] ** 2 < 10e-15,
```

Comment:
@trchudley A mysterious one that took me a bit of time to figure out! Planform and Flowline curvatures being very sensitive on flat surfaces, floating point precision started to become problematic.
For instance, for a flat DEM made only of ones `dem=np.ones((5,5), dtype=np.float32)`, where we expect 0, the Planform and Flowline would yield non-zero values (in reasonable orders of magnitude, like 0.5-2), despite all coefficients being near zero (`10e-20`)! 
This was much worse with Florinsky (due to the higher degree of derivatives) than with ZT.

I solved this by adding the `<10e-15` condition here instead of `==0`... This ideal scenario now passes in the tests.
This should hopefully suffice.